service/worker/scanner/build_ids/scavenger.go

```diff
@@ -265,9 +266,28 @@ func (a *Activities) findBuildIdsToRemove(
 	ns *namespace.Namespace,
 	entry *persistence.TaskQueueUserDataEntry,
 ) ([]string, error) {
+	// Only consider build ids that have been active at least as long as the retention time.
+	// This assumes that when a build id is added, it's used soon afterwards.
+	// This lets us avoid making visibility queries that would probably find some workflows.
+	retention := ns.Retention()
+	// Don't consider build ids that were recently the default, since there may be workers
+	// still processing tasks or data that hasn't made it to visibility yet.
+	removableBuildIdDurationSinceDefault := a.removableBuildIdDurationSinceDefault()
+
+	since := func(c *clock.HybridLogicalClock) time.Duration {
+		var t time.Time
+		if c != nil {
```

Comment:
I wish there was a way to enforce the presence of these fields in proto but since we control setting them, I think it's okay to assume their presence.

Reply:
You don't wish that :)  It causes a lot more trouble than it's worth.

I agree it's okay here but since the alternative is a NPE I wanted to be safe. It's in an activity context so it's not like it would crash the process, but still. I moved the nil check into the library so it doesn't clutter this up anymore